src/reducers/fileTree.tsx

```diff
@@ -191,6 +190,100 @@ export const getRelativePath = ({
   return pathList[newIndex];
 };
 
+const findRelativeMessageUid = (
+  currentPath: string,
+  messageMap: LinterMessageMap,
+  pathList: string[],
+  position: RelativePathPosition,
+): LinterMessage['uid'] | null => {
+  let path = currentPath;
+  for (let i = 0; i < pathList.length; i++) {
+    const nextPath = getRelativePath({
+      currentPath: path,
+      pathList,
+      position,
+    });
+    const nextMessages = messageMap[nextPath];
+    if (nextMessages) {
+      const nextMessagesForPath = getMessagesForPath(nextMessages);
+      if (position === RelativePathPosition.previous) {
+        const lastMessageIndex = nextMessagesForPath.length - 1;
+        if (lastMessageIndex >= 0) {
+          return nextMessagesForPath[lastMessageIndex].uid;
+        }
+      } else if (nextMessagesForPath.length) {
+        return nextMessagesForPath[0].uid;
+      }
+    }
+    path = nextPath;
+  }
+  // If we got here then we never found another message, which would be odd,
+  // because we should have at the very least gotten back to our original
+  // message.
+  // We should probably log this.
+  return null;
+};
+
+export type GetRelativeMessageUidParams = {
+  currentMessageUid: LinterMessage['uid'] | void;
+  currentPath: string;
+  locateCurrentMessage?: boolean;
+  messageMap: LinterMessageMap;
+  pathList: string[];
+  position: RelativePathPosition;
+};
+
+export const getRelativeMessageUid = ({
+  currentMessageUid,
+  currentPath,
+  messageMap,
+  pathList,
+  position,
+}: GetRelativeMessageUidParams): string | null => {
+  if (!currentMessageUid) {
+    // There is no current message, so get the first one.
+    if (!Object.keys(messageMap).length) {
+      return null;
+    }
+    for (const path of pathList) {
+      const messages = messageMap[path];
+      if (messages) {
+        const messagesForPath = getMessagesForPath(messages);
+        if (messagesForPath.length) {
```

Comment:
This needs a test for empty messages because I don't see a failure with this patch:

```diff
diff --git a/src/reducers/fileTree.tsx b/src/reducers/fileTree.tsx
index c956ffd..c51fb95 100644
--- a/src/reducers/fileTree.tsx
+++ b/src/reducers/fileTree.tsx
@@ -249,9 +249,7 @@ export const getRelativeMessageUid = ({
       const messages = messageMap[path];
       if (messages) {
         const messagesForPath = getMessagesForPath(messages);
-        if (messagesForPath.length) {
-          return messagesForPath[0].uid;
-        }
+        return messagesForPath[0].uid;
       }
     }
     // No messages found in the map (which shouldn't be possible because of the

```

Reply:
This is another interesting one. The `if` statement is protecting us against an empty array of messages, but I don't believe it's possible to have an empty array of messages. *If* `messages` is not `undefined`, then some messages exist in the `messageMap` and therefore the call to `getMessagesForPath` should return an array that is not empty (unless there's a problem with the data, or a bug in `getMessagesForPath`). I actually tried to write a test for this originally, but I was getting an error in my test code when I tried to do:

```
      const messageMap = createMessageMap([
        { path: file1, messages: [] },
      ]);
```

because the `createFakeLinterMessagesByPath` helper (which is used by `createMessageMap` in the test) is [designed to throw an error](https://github.com/mozilla/addons-code-manager/blob/4af1b350023ef6fa54da5ad0f0de4e4f8a23c29b/src/test-helpers.tsx#L328) if you try to use it to create some `LinterMessagesByPath` with an empty array of messages. I took this at the time to mean that this would be an invalid test, but maybe the helper needs to be changed to allow for an empty array of messages, or maybe that particular test needs to be changed to not make use of `createFakeLinterMessagesByPath`?

But I think the question of whether this check is even needed is also valid, as the app is designed to not allow this situation to exist. I think we do need it, but I just wanted to raise this as a discussion point.

Reply:
@kumar303 I'm interested to hear what you think about the above.

Reply:
You shouldn't need the helper to create a map of empty arrays. Just take a look at [the type](https://github.com/mozilla/addons-code-manager/blob/4af1b350023ef6fa54da5ad0f0de4e4f8a23c29b/src/reducers/linter.tsx#L72) and try creating it by hand.

Reply:
I believe the code has changed enough to render this comment inapplicable as well, so I'm marking it as resolved.